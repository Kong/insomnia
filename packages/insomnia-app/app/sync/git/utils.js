// @flow
export const translateSSHtoHTTP = (url: string) => {
  // handle "shorter scp-like syntax"
  url = url.replace(/^git@([^:]+):/, 'https://$1/');
  // handle proper SSH URLs
  url = url.replace(/^ssh:\/\//, 'https://');
  return url;
};

export const addDotGit = ({ url }: { url: string }) => {
  if (url.endsWith('.git')) {
    return url;
  }

  return `${url}.git`;
};

const onMessage = (message: string) => {
  console.log(`[git-event] ${message}`);
};

const onAuthFailure = (message: string) => {
  console.log(`[git-event] Auth Failure: ${message}`);
};

const onAuthSuccess = (message: string) => {
  console.log(`[git-event] Auth Success: ${message}`);
};

const onAuth = (
  credentials?: { username: string, password?: string, token?: string } = {},
) => () => ({
  username: credentials.username,
  password: credentials.password || credentials.token,
});

export const gitCallbacks = (credentials?: {
  username: string,
  password?: string,
  token?: string,
}) => ({
  onMessage,
  onAuthFailure,
  onAuthSuccess,
  onAuth: onAuth(credentials),
});
