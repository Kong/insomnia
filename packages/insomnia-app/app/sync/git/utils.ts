import type { GitCredentials } from './git-vcs';

export const translateSSHtoHTTP = (url: string) => {
  // handle "shorter scp-like syntax"
  url = url.replace(/^git@([^:]+):/, 'https://$1/');
  // handle proper SSH URLs
  url = url.replace(/^ssh:\/\//, 'https://');
  return url;
};

export const addDotGit = (url: string): string => (url.endsWith('.git') ? url : `${url}.git`);

const onMessage = (message: string) => {
  console.log(`[git-event] ${message}`);
};

const onAuthFailure = (message: string) => {
  console.log(`[git-event] Auth Failure: ${message}`);
};

const onAuthSuccess = (message: string) => {
  console.log(`[git-event] Auth Success: ${message}`);
};

// @ts-expect-error -- TSCONVERSION this needs to be handled better if credentials is undefined or which union type
const onAuth = (credentials?: GitCredentials = {}) => () => ({
  username: credentials.username,
  // @ts-expect-error -- TSCONVERSION this needs to be handled better if credentials is undefined or which union type
  password: credentials.password || credentials.token,
});

export const gitCallbacks = (credentials?: GitCredentials | null) => ({
  onMessage,
  onAuthFailure,
  onAuthSuccess,
  onAuth: onAuth(credentials || undefined),
});

// unfortunately, as of @types/node:v14.14.32 this type is not exported so we have to hackily grab it from here.
export type BufferEncoding = NonNullable<Parameters<typeof Buffer.from>[1]>;
