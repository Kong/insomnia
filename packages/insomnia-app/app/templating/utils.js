// @flow

export type NunjucksParsedTagArg = {
  type: 'string' | 'number' | 'boolean' | 'number' | 'variable' | 'expression',
  value: string | number | boolean,
  defaultValue?: string | number | boolean,
  forceVariable?: boolean,
  placeholder?: string,
  help?: string,
  displayName?: string,
  quotedBy?: '"' | "'",
  validate?: (value: any) => string,
  hide?: (Array<NunjucksParsedTagArg>) => boolean,
};

export type NunjucksParsedTag = {
  name: string,
  args: Array<NunjucksParsedTagArg>,
  rawValue?: string,
};

/**
 * Get list of paths to all primitive types in nested object
 * @param {object} obj - object to analyse
 * @param {String} [prefix] - base path to prefix to all paths
 * @returns {Array} - list of paths
 */
export function getKeys(obj: any, prefix: string = ''): Array<{ name: string, value: any }> {
  let allKeys = [];

  const typeOfObj = Object.prototype.toString.call(obj);

  if (typeOfObj === '[object Array]') {
    for (let i = 0; i < obj.length; i++) {
      allKeys = [...allKeys, ...getKeys(obj[i], `${prefix}[${i}]`)];
    }
  } else if (typeOfObj === '[object Object]') {
    const keys = Object.keys(obj);
    for (const key of keys) {
      const newPrefix = prefix ? `${prefix}.${key}` : key;
      allKeys = [...allKeys, ...getKeys(obj[key], newPrefix)];
    }
  } else if (typeOfObj === '[object Function]') {
    // Ignore functions
  } else if (prefix) {
    allKeys.push({ name: prefix, value: obj });
  }

  return allKeys;
}

/**
 * Parse a Nunjucks tag string into a usable abject
 * @param {string} tagStr - the template string for the tag
 * @return {object} parsed tag data
 */
export function tokenizeTag(tagStr: string): NunjucksParsedTag {
  // ~~~~~~~~ //
  // Sanitize //
  // ~~~~~~~~ //

  const withoutEnds = tagStr
    .trim()
    .replace(/^{%/, '')
    .replace(/%}$/, '')
    .trim();

  const nameMatch = withoutEnds.match(/^[a-zA-Z_$][0-9a-zA-Z_$]*/);
  const name = nameMatch ? nameMatch[0] : withoutEnds;
  const argsStr = withoutEnds.slice(name.length);

  // ~~~~~~~~~~~~~ //
  // Tokenize Args //
  // ~~~~~~~~~~~~~ //

  const args = [];
  let quotedBy = null;
  let currentArg = null;
  for (let i = 0; i < argsStr.length + 1; i++) {
    // Adding an "invisible" at the end helps us terminate the last arg
    const c = argsStr.charAt(i) || ',';

    // Do nothing if we're still on a space o comma
    if (currentArg === null && c.match(/[\s,]/)) {
      continue;
    }

    // Start a new single-quoted string
    if (currentArg === null && c === "'") {
      currentArg = '';
      quotedBy = "'";
      continue;
    }

    // Start a new double-quoted string
    if (currentArg === null && c === '"') {
      currentArg = '';
      quotedBy = '"';
      continue;
    }

    // Start a new unquoted string
    if (currentArg === null) {
      currentArg = c;
      quotedBy = null;
      continue;
    }

    const endQuoted = quotedBy && c === quotedBy;
    const endUnquoted = !quotedBy && c === ',';
    const argCompleted = endQuoted || endUnquoted;

    // Append current char to argument
    if (!argCompleted && currentArg !== null) {
      if (c === '\\') {
        // Handle backslashes
        i += 1;
        currentArg += argsStr.charAt(i);
      } else {
        currentArg += c;
      }
    }

    // End current argument
    if (currentArg !== null && argCompleted) {
      let arg;
      if (quotedBy) {
        arg = { type: 'string', value: currentArg, quotedBy };
      } else if (['true', 'false'].includes(currentArg)) {
        arg = { type: 'boolean', value: currentArg.toLowerCase() === 'true' };
      } else if (currentArg.match(/^\d*\.?\d*$/)) {
        arg = { type: 'number', value: currentArg };
      } else if (currentArg.match(/^[a-zA-Z_$][0-9a-zA-Z_$]*$/)) {
        arg = { type: 'variable', value: currentArg };
      } else {
        arg = { type: 'expression', value: currentArg };
      }

      args.push(arg);

      currentArg = null;
      quotedBy = null;
    }
  }

  return { name, args };
}

/** Convert a tokenized tag back into a Nunjucks string */
export function unTokenizeTag(tagData: NunjucksParsedTag): string {
  const args = [];
  for (const arg of tagData.args) {
    if (['string', 'model', 'file'].includes(arg.type)) {
      const q = arg.quotedBy || "'";
      const re = new RegExp(`([^\\\\])${q}`, 'g');
      const str = arg.value.toString().replace(re, `$1\\${q}`);
      args.push(`${q}${str}${q}`);
    } else if (arg.type === 'boolean') {
      args.push(arg.value ? 'true' : 'false');
    } else {
      args.push(arg.value);
    }
  }

  const argsStr = args.join(', ');
  return `{% ${tagData.name} ${argsStr} %}`;
}

/** Get the default Nunjucks string for an extension */
export function getDefaultFill(name: string, args: Array<NunjucksParsedTagArg>): string {
  const stringArgs: Array<string> = (args || []).map(argDefinition => {
    switch (argDefinition.type) {
      case 'enum':
        const { defaultValue, options } = argDefinition;
        const value = defaultValue !== undefined ? defaultValue : options[0].value;
        return `'${value}'`;
      case 'number':
        return `${parseFloat(argDefinition.defaultValue) || 0}`;
      case 'boolean':
        return argDefinition.defaultValue ? 'true' : 'false';
      case 'string':
      case 'file':
      case 'model':
        return `'${(argDefinition.defaultValue: any) || ''}'`;
      default:
        return "''";
    }
  });

  return `${name} ${stringArgs.join(', ')}`;
}
