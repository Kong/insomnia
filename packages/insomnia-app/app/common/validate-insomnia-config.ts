import { getConfigSettings, isConfigError, isParseError } from '../models/helpers/settings';

interface Result {
  error?:{
    title: string;
    message: string;
  };
}

export const validateInsomniaConfig = (): Result => {
  const configSettings = getConfigSettings();

  if (!('error' in configSettings)) {
    return {};
  }

  let title = 'Invalid Insomnia Config';
  let message = '';

  if (isParseError(configSettings)) {
    const { syntaxError, configPath } = configSettings.error;
    message = [
      'Failed to parse JSON file for Insomnia Config.',
      '',
      '[Path]',
      configPath,
      '',
      '[Syntax Error]',
      syntaxError.message,
    ].join('\n');
  } else if (isConfigError(configSettings)) {
    const { humanReadableErrors, configPath } = configSettings.error;
    const errors = humanReadableErrors.map(({ message, path, suggestion }, index) => ([
      `[Error ${index + 1}]`,
      `Path: ${path}`,
      `${message}.${suggestion ? `  ${suggestion}` : ''}`,
    ]).join('\n')).join('\n\n');

    message = [
      `Your Insomnia Config was found to be invalid.  Please check the path below for the following error${configSettings.error.humanReadableErrors?.length > 1 ? 's' : ''}:`,
      '',
      '[Path]',
      configPath,
      '',
      errors,
    ].join('\n');
  } else {
    title = 'An unexpected error occured while parsing Insomnia Config';
    message = JSON.stringify(configSettings);
  }

  return { error: { title, message } };
};
