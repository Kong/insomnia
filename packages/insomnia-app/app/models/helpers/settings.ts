import { readFileSync } from 'fs';
import { Settings } from 'insomnia-common';
import { ErrorResult, INSOMNIA_CONFIG_FILENAME, InsomniaConfig, isErrorResult, validate } from 'insomnia-config';
import { resolve } from 'path';
import { mapObjIndexed, once } from 'ramda';
import { omitBy } from 'ramda-adjunct';
import { ValueOf } from 'type-fest';

import { isDevelopment } from '../../common/constants';
import { getDataDirectory, getPortableExecutableDir } from '../../common/electron-helpers';

interface FailedParseResult {
  syntaxError: SyntaxError;
  fileContents: string;
  configPath: string;
}

const isFailedParseResult = (input: any): input is FailedParseResult => {
  const typesafeInput = input as FailedParseResult;

  return (
    typesafeInput ? typesafeInput.syntaxError instanceof SyntaxError : false
  );
};

/** takes an unresolved (or resolved will work fine too) filePath of the insomnia config and reads the insomniaConfig from disk */
export const readConfigFile = (configPath?: string): unknown | FailedParseResult | undefined => {
  if (!configPath) {
    return undefined;
  }

  let fileContents = '';
  try {
    fileContents = readFileSync(configPath, 'utf-8');
  } catch (error: unknown) {
    // file not found
    return undefined;
  }

  const fileIsFoundButEmpty = fileContents === '';
  if (fileIsFoundButEmpty) {
    return undefined;
  }

  try {
    return JSON.parse(fileContents) as unknown;
  } catch (syntaxError: unknown) {
    // note: all JSON.parse errors are SyntaxErrors
    // see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/JSON_bad_parse
    if (syntaxError instanceof SyntaxError) {
      console.error('failed to parse insomnia config', { configPath, fileContents, syntaxError });
      const failedParseResult: FailedParseResult = {
        syntaxError,
        fileContents,
        configPath: configPath,
      };
      return failedParseResult;
    }
    return undefined;
  }
};

export const getLocalDevConfigFilePath = () => (
  isDevelopment() ? '../../packages/insomnia-app/app' as string : undefined
);

const addConfigFileToPath = (path: string | undefined) => (
  path ? resolve(path, INSOMNIA_CONFIG_FILENAME) : undefined
);

export const getConfigFile = () => {
  const portableExecutable = getPortableExecutableDir();
  const insomniaDataDirectory = getDataDirectory();
  const localDev = getLocalDevConfigFilePath();
  const configPaths = [
    portableExecutable,
    insomniaDataDirectory,
    localDev,
  ].map(addConfigFileToPath);

  // note: this is written as to avoid unnecessary (synchronous) reads from disk.
  // The paths above are in priority order such that if we already found what we're looking for, there's no reason to keep reading other files.
  for (const configPath of configPaths) {
    const fileReadResult = readConfigFile(configPath);
    if (isFailedParseResult(fileReadResult)) {
      return fileReadResult;
    }

    if (fileReadResult !== undefined && configPath !== undefined) {
      return {
        insomniaConfig: fileReadResult,
        configPath,
      };
    }
  }
  const fallbackEmptyConfig: InsomniaConfig = { insomniaConfig: '1.0.0' };
  return {
    insomniaConfig: fallbackEmptyConfig,
    configPath: '<internal fallback insomnia config>',
  };
};

export interface ConfigError {
  error: {
    configPath?: string;
    insomniaConfig: unknown;
    errors: ErrorResult['errors'];
    humanReadableErrors: ErrorResult['humanReadableErrors'];
  };
}

export const isConfigError = (input: ConfigError | ParseError): input is ConfigError => (
  // Cast for typesafety
  (input as ConfigError).error?.humanReadableErrors?.length > 0
);

export interface ParseError {
  error: FailedParseResult;
}

export const isParseError = (input: ConfigError | ParseError): input is ParseError => (
  isFailedParseResult(input.error)
);

/**
 * gets settings from the `insomnia.config.json`
 *
 * note that it is a business rule that the config is never read again after startup, hence the `once` usage.
 */
export const getConfigSettings: () => (NonNullable<InsomniaConfig['settings']> | ConfigError | ParseError) = once(() => {
  const configFileResult = getConfigFile();

  if (isFailedParseResult(configFileResult)) {
    return {
      error: configFileResult,
    };
  }

  const { insomniaConfig, configPath } = configFileResult;
  const validationResult = validate(insomniaConfig as InsomniaConfig);

  if (isErrorResult(validationResult)) {
    const { errors, humanReadableErrors } = validationResult;
    const error = {
      configPath,
      insomniaConfig,
      errors,
      humanReadableErrors,
    };

    console.error('invalid insomnia config', error);
    return { error };
  }

  // This cast is important for testing intentionally bad values (the above validation will catch it, anyway)
  return (insomniaConfig as InsomniaConfig).settings || {};
});

interface Condition {
  /** note: conditions are only suitable for boolean settings at this time */
  when: boolean;
  set: Partial<Settings>;
}

// using a Map because they are ordered (in such case as multiple settings could be controlled by other multiple settings in the future, we could control this reliably by changing the order in the Map)
const settingControllers = new Map<keyof Settings, Condition[]>([
  [
    'incognitoMode',
    [
      {
        when: true,
        set: {
          enableAnalytics: false,
          allowNotificationRequests: false,
        },
      },
    ],
  ],
]);

/** checks whether a given setting is literally specified in the insomnia config */
export const isControlledByConfig = (setting: keyof Settings | null) => setting ? (
  Boolean(Object.prototype.hasOwnProperty.call(getConfigSettings(), setting))
) : false;

export interface SettingControlledSetting<T extends keyof Settings> {
  controlledValue: Settings[T];
  controller: keyof Settings;
  isControlled: true;
}

export interface ConfigControlledSetting<T extends keyof Settings> {
  controlledValue: Settings[T];
  controller: 'insomnia-config';
  isControlled: true;
}

export interface UncontrolledSetting {
  controller: null;
  isControlled: false;
}

export type SettingsControl<T extends keyof Settings> =
  | SettingControlledSetting<T>
  | ConfigControlledSetting<T>
  | UncontrolledSetting
;

const isSettingControlledByCondition = (condition: Condition, setting: keyof Settings, value: ValueOf<Settings>) => {
  return condition.when === value
    && Object.prototype.hasOwnProperty.call(condition.set, setting);
};

/**
 * checks whether a given setting is controlled by another setting.
 * if so, it will return that setting id.  otherwise it will return false.
 */
export const isControlledByAnotherSetting = (settings: Settings) => (setting: keyof Settings) => {
  for (const [controller, controlledSettings] of settingControllers.entries()) {
    for (const condition of controlledSettings) {
      if (isSettingControlledByCondition(condition, setting, settings[controller])) {
        const output: SettingControlledSetting<typeof setting> = {
          controlledValue: condition.set[setting],
          controller,
          isControlled: true,
        };
        return output;
      }
    }
  }
  const uncontrolledSetting: UncontrolledSetting = {
    controller: null,
    isControlled: false,
  };
  return uncontrolledSetting;
};

/**
 * For any given setting, return what the value of that setting should be once you take the insomnia config and other potentially controlling settings into account
 */
export const getControlledStatus = (userSettings: Settings) => (setting: keyof Settings) => {
  const configSettings = {
    ...userSettings,
    ...getConfigSettings(),
  };

  if (isControlledByConfig(setting)) {

    // note that the raw config settings are being passed here (rathern than `settings` alone), because we must verify that the controller does not itself also have a specification in the config
    const controllerSetting = isControlledByAnotherSetting(configSettings)(setting);

    // TLDR; the config always wins
    // It is a business rule that if a setting (e.g. `incognitoMode` specified in the config controlls another setting (e.g. `enableAnalytics`), that conflict should be resolved such that the config-specified controller should _always_ win, _even_ if the controlled setting (i.e. `enableAnalytics`) is _itself_ specified in the config with a conflicting value)
    if (controllerSetting.isControlled && isControlledByConfig(controllerSetting.controller)) {
      // since this setting is also controlled by a controller, and that controller is controlled by the config, we use the controller's desired value for this setting.
      return {
        controller: controllerSetting.controller,
        isControlled: true,
        value: controllerSetting.controlledValue,
      };
    }

    // no other setting controls this, so we can grab its value the config directly
    return {
      controller: 'insomnia-config',
      isControlled: true,
      value: configSettings[setting],
    };
  }

  const thisSetting = isControlledByAnotherSetting(configSettings)(setting);
  if (thisSetting.isControlled) {
    // this setting is controlled by another setting.
    return {
      controller: thisSetting.controller,
      isControlled: true,
      value: thisSetting.controlledValue,
    };
  }

  // return the object unchanged, as it exists in the settings
  return {
    controller: null,
    isControlled: false,
    value: userSettings[setting],
  };
};

/** removes any setting in the given patch object which is controlled in any way (i.e. either by the insomnia config or by another setting) */
export const omitControlledSettings = <
  T extends Settings,
  U extends Partial<Settings>
>(settings: T, patch: U) => {
  return omitBy((_value, setting: keyof Settings) => (
    getControlledStatus(settings)(setting).isControlled
  ), patch);
};

/** for any given setting, whether controlled by the insomnia config or whether controlled by another value, return the calculated value */
export const getMonkeyPatchedControlledSettings = <T extends Settings>(settings: T) => {
  const override = mapObjIndexed((_value, setting: keyof Settings) => (
    getControlledStatus(settings)(setting).value
  ), settings) as T;
  return {
    ...settings,
    ...override,
  };
};
