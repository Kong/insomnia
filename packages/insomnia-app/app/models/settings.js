// @flow
import type { BaseModel } from './index';
import * as db from '../common/database';
import { UPDATE_CHANNEL_STABLE } from '../common/constants';

type BaseSettings = {
  showPasswords: boolean,
  useBulkHeaderEditor: boolean,
  followRedirects: boolean,
  editorFontSize: number,
  editorIndentSize: number,
  editorLineWrapping: boolean,
  editorKeyMap: string,
  editorIndentWithTabs: boolean,
  httpProxy: string,
  httpsProxy: string,
  noProxy: string,
  proxyEnabled: boolean,
  timeout: number,
  validateSSL: boolean,
  forceVerticalLayout: boolean,
  autoHideMenuBar: boolean,
  theme: string,
  maxRedirects: number,
  pluginPath: string,
  nunjucksPowerUserMode: boolean,
  deviceId: string | null,
  updateChannel: string,
  updateAutomatically: boolean,
  disableUpdateNotification: boolean,
  environmentHighlightColorStyle: string,
  fontMonospace: string | null,
  fontInterface: string | null,
  fontSize: number,
  fontVariantLigatures: boolean
};

export type Settings = BaseModel & BaseSettings;

export const name = 'Settings';
export const type = 'Settings';
export const prefix = 'set';
export const canDuplicate = false;

export function init(): BaseSettings {
  return {
    showPasswords: false,
    useBulkHeaderEditor: false,
    followRedirects: true,
    editorFontSize: 11,
    editorIndentSize: 2,
    editorLineWrapping: true,
    editorKeyMap: 'default',
    editorIndentWithTabs: true,
    httpProxy: '',
    httpsProxy: '',
    noProxy: '',
    maxRedirects: -1,
    proxyEnabled: false,
    timeout: 0,
    validateSSL: true,
    forceVerticalLayout: false,
    autoHideMenuBar: false,
    theme: 'default',
    pluginPath: '',
    nunjucksPowerUserMode: false,
    deviceId: null,
    updateChannel: UPDATE_CHANNEL_STABLE,
    updateAutomatically: true,
    disableUpdateNotification: false,
    environmentHighlightColorStyle: 'sidebar-indicator',
    fontMonospace: null,
    fontInterface: null,
    fontSize: 13,
    fontVariantLigatures: false
  };
}

export function migrate(doc: Settings): Settings {
  return doc;
}

export async function all(patch: Object = {}): Promise<Array<Settings>> {
  const settings = await db.all(type);
  if (settings.length === 0) {
    return [await getOrCreate()];
  } else {
    return settings;
  }
}

export async function create(patch: Object = {}): Promise<Settings> {
  return db.docCreate(type, patch);
}

export async function update(settings: Settings, patch: Object): Promise<Settings> {
  return db.docUpdate(settings, patch);
}

export async function getOrCreate(patch: Object = {}): Promise<Settings> {
  const results = await db.all(type);
  if (results.length === 0) {
    return create(patch);
  } else {
    return results[0];
  }
}
