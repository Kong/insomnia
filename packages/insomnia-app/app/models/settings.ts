import type { BaseModel } from './index';
import { database as db } from '../common/database';
import {
  getAppDefaultTheme,
  getAppDefaultLightTheme,
  getAppDefaultDarkTheme,
  HttpVersions,
  UPDATE_CHANNEL_STABLE,
} from '../common/constants';
import * as hotkeys from '../common/hotkeys';
import type { HttpVersion } from '../common/constants';

export interface PluginConfig {
  disabled: boolean;
}

export type PluginConfigMap = Record<string, PluginConfig>;

interface BaseSettings {
  autoHideMenuBar: boolean;
  autocompleteDelay: number;
  deviceId: string | null;
  disableHtmlPreviewJs: boolean;
  disableResponsePreviewLinks: boolean;
  disableUpdateNotification: boolean;
  editorFontSize: number;
  editorIndentSize: number;
  editorIndentWithTabs: boolean;
  editorKeyMap: string;
  editorLineWrapping: boolean;
  enableAnalytics: boolean;
  environmentHighlightColorStyle: string;
  filterResponsesByEnv: boolean;
  followRedirects: boolean;
  clearOAuth2SessionOnRestart: boolean;
  fontInterface: string | null;
  fontMonospace: string | null;
  fontSize: number;
  fontVariantLigatures: boolean;
  forceVerticalLayout: boolean;
  hotKeyRegistry: hotkeys.HotKeyRegistry;
  httpProxy: string;
  httpsProxy: string;
  lineWrapping?: boolean;
  maxHistoryResponses: number;
  maxRedirects: number;
  maxTimelineDataSizeKB: number;
  noProxy: string;
  nunjucksPowerUserMode: boolean;
  pluginConfig: PluginConfigMap;
  pluginPath: string;
  preferredHttpVersion: HttpVersion;
  proxyEnabled: boolean;
  showPasswords: boolean;
  theme: string;
  autoDetectColorScheme: boolean;
  lightTheme: string;
  darkTheme: string;
  timeout: number;
  updateAutomatically: boolean;
  updateChannel: string;
  useBulkHeaderEditor: boolean;
  useBulkParametersEditor: boolean;
  validateSSL: boolean;
  hasPromptedToMigrateFromDesigner: boolean;
  hasPromptedOnboarding: boolean;
  hasPromptedAnalytics: boolean;
}

export type Settings = BaseModel & BaseSettings;
export const name = 'Settings';
export const type = 'Settings';
export const prefix = 'set';
export const canDuplicate = false;
export const canSync = false;

export type ThemeSettings = Pick<Settings, 'autoDetectColorScheme' | 'lightTheme' | 'darkTheme' | 'theme'>;

export const isSettings = (model: Pick<BaseModel, 'type'>): model is Settings => (
  model.type === type
);

export function init(): BaseSettings {
  return {
    autoHideMenuBar: false,
    autocompleteDelay: 1200,
    deviceId: null,
    disableHtmlPreviewJs: false,
    disableResponsePreviewLinks: false,
    disableUpdateNotification: false,
    editorFontSize: 11,
    editorIndentSize: 2,
    editorIndentWithTabs: true,
    editorKeyMap: 'default',
    editorLineWrapping: true,
    enableAnalytics: false,
    environmentHighlightColorStyle: 'sidebar-indicator',
    filterResponsesByEnv: false,
    followRedirects: true,
    clearOAuth2SessionOnRestart: true,
    fontInterface: null,
    fontMonospace: null,
    fontSize: 13,
    fontVariantLigatures: false,
    forceVerticalLayout: false,
    hotKeyRegistry: hotkeys.newDefaultRegistry(),
    httpProxy: '',
    httpsProxy: '',
    maxHistoryResponses: 20,
    maxRedirects: -1,
    maxTimelineDataSizeKB: 10,
    noProxy: '',
    nunjucksPowerUserMode: false,
    pluginConfig: {},
    pluginPath: '',
    preferredHttpVersion: HttpVersions.default,
    proxyEnabled: false,
    showPasswords: false,
    theme: getAppDefaultTheme(),
    autoDetectColorScheme: false,
    lightTheme: getAppDefaultLightTheme(),
    darkTheme: getAppDefaultDarkTheme(),
    timeout: 0,
    updateAutomatically: true,
    updateChannel: UPDATE_CHANNEL_STABLE,
    useBulkHeaderEditor: false,
    useBulkParametersEditor: false,
    validateSSL: true,
    hasPromptedToMigrateFromDesigner: false,
    // Users should only see onboarding during first launch, and anybody updating from an
    // older version should not see it, so by default this flag is set to true, and is toggled
    // to false during initialization
    hasPromptedOnboarding: true,
    // Only existing users updating from an older version should see the analytics prompt
    // So by default this flag is set to false, and is toggled to true during initialization
    // for new users
    hasPromptedAnalytics: false,
  };
}

export function migrate(doc: Settings) {
  doc = migrateEnsureHotKeys(doc);
  return doc;
}

export async function all() {
  const settings = await db.all<Settings>(type);

  if (settings?.length === 0) {
    return [await getOrCreate()];
  } else {
    return settings;
  }
}

export async function create(patch: Partial<Settings> = {}) {
  return db.docCreate<Settings>(type, patch);
}

export async function update(settings: Settings, patch: Partial<Settings>) {
  return db.docUpdate<Settings>(settings, patch);
}

export async function patch(patch: Partial<Settings>) {
  const settings = await getOrCreate();
  return db.docUpdate<Settings>(settings, patch);
}

export async function getOrCreate() {
  const results = await db.all<Settings>(type) || [];

  if (results.length === 0) {
    return create();
  } else {
    return results[0];
  }
}

/**
 * Ensure map is updated when new hotkeys are added
 */
function migrateEnsureHotKeys(settings: Settings): Settings {
  settings.hotKeyRegistry = { ...hotkeys.newDefaultRegistry(), ...settings.hotKeyRegistry };
  return settings;
}
