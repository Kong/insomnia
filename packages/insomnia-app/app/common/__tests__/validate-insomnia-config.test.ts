import { mocked } from 'ts-jest/utils';

import { ConfigError, getConfigSettings as _getConfigSettings, ParseError  } from '../../models/helpers/settings';
import { validateInsomniaConfig } from '../validate-insomnia-config';

jest.mock('../../models/helpers/settings');
const getConfigSettings = mocked(_getConfigSettings);

describe('validateInsomniaConfig', () => {
  it('should show error box and exit if there is a parse error', () => {
    // Arrange
    const errorReturn: ParseError = {
      error: {
        syntaxError: new SyntaxError('mock syntax error'),
        fileContents: '{ "mock": ["insomnia", "config"] }',
        configPath: '/mock/insomnia/config/path',
      },
    };
    getConfigSettings.mockReturnValue(errorReturn);

    // Act
    const result = validateInsomniaConfig();

    // Assert
    expect(result.error?.title).toBe('Invalid Insomnia Config');
    expect(result.error?.message).toMatchSnapshot();
  });

  it('should show error box and exit if there is a config error', () => {
    // Arrange
    const errorReturn: ConfigError = {
      error: {
        errors: [],
        humanReadableErrors: [{
          message: 'message',
          path: 'path',
          suggestion: 'suggestion',
          context: { errorType: 'const' },
        }],
        insomniaConfig: '{ "mock": ["insomnia", "config"] }',
        configPath: '/mock/insomnia/config/path',
      },
    };
    getConfigSettings.mockReturnValue(errorReturn);

    // Act
    const result = validateInsomniaConfig();

    // Assert
    expect(result.error?.title).toBe('Invalid Insomnia Config');
    expect(result.error?.message).toMatchSnapshot();
  });

  it('should show error box and exit if there is an unexpected error return', () => {
    // Arrange
    const errorReturn: ConfigError = {
      error: {
        errors: [],
        humanReadableErrors: [],
        insomniaConfig: '{ "mock": ["insomnia", "config"] }',
        configPath: '/mock/insomnia/config/path',
      },
    };
    getConfigSettings.mockReturnValue(errorReturn);

    // Act
    const result = validateInsomniaConfig();

    // Assert
    expect(result.error?.title).toBe('An unexpected error occured while parsing Insomnia Config');
    expect(result.error?.message).toMatchSnapshot();
  });

  it('should not exit if there are no errors', () => {
    // Arrange
    const validReturn = { enableAnalytics: true };
    getConfigSettings.mockReturnValue(validReturn);

    // Act
    const result = validateInsomniaConfig();

    // Assert
    expect(result.error).not.toBeDefined();
  });
});
