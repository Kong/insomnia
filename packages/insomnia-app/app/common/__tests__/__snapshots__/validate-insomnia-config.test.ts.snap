// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateInsomniaConfig should return error if there is a config error 1`] = `"{\\"error\\":{\\"errors\\":[],\\"humanReadableErrors\\":[],\\"insomniaConfig\\":\\"{ \\\\\\"mock\\\\\\": [\\\\\\"insomnia\\\\\\", \\\\\\"config\\\\\\"] }\\",\\"configPath\\":\\"/mock/insomnia/config/path\\"}}"`;

exports[`validateInsomniaConfig should return error if there is a parse error 1`] = `
"Failed to parse JSON file for Insomnia Config.

[Path]
/mock/insomnia/config/path

[Syntax Error]
mock syntax error"
`;
